File: backend/guardian-boarding/src/main/java/com/guardian/boarding/domain/HandoverCode.java
package com.guardian.boarding.domain;

import java.time.Instant;
import java.util.Objects;
import java.util.UUID;

/**
 * A short-lived verification code a guardian shows to release their child at drop (MOD-09, screen
 * P-12).
 *
 * <p>This is the QR / numeric-PIN method BR-HAND-002 requires at least one of. The code is shown to
 * the attendant; it does not, by itself, release anyone — that verification step lives in the
 * driver/attendant app, which does not yet exist in this codebase. See {@code
 * guardian-boarding/build.gradle.kts} for the fuller scope note.
 *
 * @param code exactly six digits, matching the mockup's grouped display ("4 8 2 9 1 3").
 */
public record HandoverCode(
    UUID id,
    UUID studentId,
    UUID requestedByGuardianId,
    String code,
    Instant issuedAt,
    Instant expiresAt) {

  public HandoverCode {
    Objects.requireNonNull(studentId, "studentId");
    Objects.requireNonNull(code, "code");
    Objects.requireNonNull(issuedAt, "issuedAt");
    Objects.requireNonNull(expiresAt, "expiresAt");

    if (!code.matches("\\d{6}")) {
      throw new IllegalArgumentException("code must be exactly six digits");
    }
    if (!expiresAt.isAfter(issuedAt)) {
      throw new IllegalArgumentException("expiresAt must be after issuedAt");
    }
  }

  /** Seconds remaining until this code stops being shown as live, floored at zero. */
  public long secondsRemaining(Instant now) {
    long seconds = expiresAt.getEpochSecond() - now.getEpochSecond();
    return Math.max(0, seconds);
  }
}

File: flutter/parent_app/lib/features/handover/repository/models/handover_code.dart
import 'package:flutter/foundation.dart';

/// A short-lived verification code shown to the bus attendant at drop — P-12.
///
/// This is what the parent shows; it is not itself a handover. Verifying and redeeming it at
/// the vehicle is the attendant-side flow (BR-HAND-001 through BR-HAND-007), which this app
/// does not implement.
@immutable
class HandoverCode {
  const HandoverCode({
    required this.studentId,
    required this.code,
    required this.expiresAt,
  });

  final String studentId;

  /// Exactly six digits. Also the QR payload — a second, independent secret would be one
  /// more thing that can fail to match at the vehicle.
  final String code;

  final DateTime expiresAt;

  bool get isExpired => DateTime.now().isAfter(expiresAt);

  Duration remaining([DateTime? now]) {
    final diff = expiresAt.difference(now ?? DateTime.now());
    return diff.isNegative ? Duration.zero : diff;
  }

  /// The mockup's grouped display: "4 8 2 9 1 3".
  String get grouped => code.split('').join(' ');
}
